extensions/data-plane/data-plane-google-bigquery/src/main/java/org/eclipse/edc/connector/dataplane/gcp/bigquery/pipeline/BigQueryDataSink.java

```diff
@@ -0,0 +1,337 @@
+ */
+public class BigQueryDataSink extends ParallelSink {
+    private BigQueryConfiguration configuration;
+    private BigQueryTarget target;
+    private GoogleCredentials credentials;
+    private final Phaser inflightRequestCount = new Phaser(1);
+    private BigQueryWriteClient writeClient;
+    private JsonStreamWriter streamWriter;
+
+    private BigQueryDataSink() {
+    }
+
+    void testAppendSignal() {
+        inflightRequestCount.arriveAndDeregister();
+    }
+
+    @Override
+    protected StreamResult<Object> transferParts(List<DataSource.Part> parts) {
+        if (parts.isEmpty()) {
+            return StreamResult.success();
+        }
+
+        Exception appendException = null;
+        var errorWhileAppending = false;
+        for (var part : parts) {
+            try (var inputStream = part.openStream()) {
+                var mapper = new ObjectMapper();
+                var jsonParser = mapper.createParser(inputStream);
+                var token = jsonParser.nextToken();
+                while (token != null && token == JsonToken.START_ARRAY) {
+                    var element = mapper.readTree(jsonParser);
+                    var jsonString = element.toString();
+                    var page = new JSONArray(jsonString);
+                    append(page);
+                    token = jsonParser.nextToken();
+                }
+
+                if (token != null && token != JsonToken.START_ARRAY) {
+                    throw new IllegalStateException("BigQuery sink JSON array expected");
+                }
+            } catch (Exception exception) {
+                errorWhileAppending = true;
+                appendException = exception;
+                monitor.severe("BigQuery Sink error while appending: ", appendException);
+                break;
+            }
+
+            if (part instanceof BigQueryPart bigQueryPart) {
+                if (bigQueryPart.getException() != null) {
+                    errorWhileAppending = true;
+                    appendException = bigQueryPart.getException();
+                    monitor.severe("BigQuery Sink error from source: ", appendException);
+                    break;
+                }
+            }
+        }
+
+        inflightRequestCount.arriveAndAwaitAdvance();
+        closeSinkStream();
+
+        monitor.debug("BigQuery Sink requests arrived, stream " + streamWriter.getStreamName() + " for table " + target.getTableName().toString() + " closed, now finalizing..");
+        var finalizeResponse =
+                writeClient.finalizeWriteStream(streamWriter.getStreamName());
+        monitor.debug("BigQuery Sink rows written: " + finalizeResponse.getRowCount());
+
+        // Commit the data stream received only if the stream writer succeeded (transaction).
+        // TODO support option to enable commit even if the stream writer failed.
+        if (!errorWhileAppending) {
+            commitData();
+        }
+
+        writeClient.shutdownNow();
```

Comment:
personally, i would refactor L130 - 145 in a separate method, maybe even L116-145. Something like a `cleanupResources()` or similar.

Reply:
Done